### ISSUE TITLE
T-SQL: HASH JOIN not supported

### ISSUE THREAD

**Issue by borchero** [sqlfluff/sqlfluff#3916]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

`sqlfluff` does not recognize (i.e. cannot parse) the T-SQL statement `INNER HASH JOIN`.

### Expected Behaviour

[Join hints](https://learn.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-join?view=sql-server-ver16) should be recognized by `sqlfluff`.

### Observed Behaviour

```
== [test.sql] FAIL                                                                                                                      
L:   3 | P:   1 |  PRS | Line 3, Position 1: Found unparsable section: 'INNER                                                           
                       | HASH JOIN b\n ON a.col = b.col\n'
WARNING: Parsing errors found and dialect is set to 'tsql'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

```sql
SELECT a.*
FROM a
INNER HASH JOIN b
    ON a.col = b.col
```

### Dialect

`tsql`

### Version

- sqlfluff, version 1.3.1
- Installed via homebrew

### Configuration

```
[sqlfluff]
dialect = tsql
exclude_rules = L011,L012,L013,L031,L034,L044
large_file_skip_byte_limit = 0

[sqlfluff:indentation]
indented_ctes = true
indented_joins = false

[sqlfluff:rules]
comma_style = trailing
group_by_and_order_by_style = explicit
max_line_length = 99
tab_space_size = 4

[sqlfluff:rules:L010]
capitalisation_policy = upper
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
